substrate-node/pallets/pallet-smart-contract/src/benchmarking.rs

```diff
@@ -1,207 +1,532 @@
+        caller.clone()),
+        b"some_description".to_vec(),
+        b"some_link".to_vec(),
+        providers.clone()
+    )
     verify {
-        let contract = SmartContractModule::<T>::contracts(1).unwrap();
+        let solution_provider_id = 1;
+        assert!(SmartContractModule::<T>::solution_providers(solution_provider_id).is_some());
+        let solution_provider = SmartContractModule::<T>::solution_providers(solution_provider_id).unwrap();
         assert_eq!(
-            contract.contract_id, 1
+            solution_provider.providers, providers
         );
+        assert_last_event::<T>(Event::SolutionProviderCreated(solution_provider).into());
     }
-}
 
-pub fn create_twin<T: Config>(source: T::AccountId) {
-    assert_ok!(pallet_tfgrid::Pallet::<T>::user_accept_tc(
-        RawOrigin::Signed(source.clone()).into(),
-        get_document_link_input(b"some_link"),
-        get_document_hash_input(b"some_hash"),
-    ));
+    // approve_solution_provider()
+    approve_solution_provider {
+        let provider: T::AccountId = account("Alice", 0, 0);
+        _create_twin::<T>(provider.clone());
+        _create_solution_provider::<T>(provider.clone());
+        let solution_provider_id = 1;
+        let approve = true;
 
-    let relay = get_relay_input(b"somerelay.io");
-    let pk =
-        get_public_key_input(b"0x6c8fd181adc178cea218e168e8549f0b0ff30627c879db9eac4318927e87c901");
-    assert_ok!(pallet_tfgrid::Pallet::<T>::create_twin(
-        RawOrigin::Signed(source).into(),
-        relay,
-        pk
-    ));
-}
+    }: _(RawOrigin::Root, solution_provider_id, approve)
+    verify {
+        assert!(SmartContractModule::<T>::solution_providers(solution_provider_id).is_some());
+        let solution_provider = SmartContractModule::<T>::solution_providers(solution_provider_id).unwrap();
+        assert_last_event::<T>(Event::SolutionProviderApproved(solution_provider_id, approve).into());
+    }
 
-pub fn prepare_farm_and_node<T: Config>(source: T::AccountId) {
-    create_twin::<T>(source.clone());
-    prepare_farm::<T>(source.clone());
+    // bill_contract_for_block()
+    bill_contract_for_block {
+        let farmer: T::AccountId = account("Alice", 0, 0);
+        _prepare_farm_with_node::<T>(farmer.clone());
+
+        let user: T::AccountId = account("Bob", 0, 1);
+        let user_lookup = T::Lookup::unlookup(user.clone());
+        let balance_init_amount = <T as pallet_balances::Config>::Balance::saturated_from(100000000 as u128);
+        Balances::<T>::force_set_balance(RawOrigin::Root.into(), user_lookup, balance_init_amount).unwrap();
+        _create_twin::<T>(user.clone());
+        _create_node_contract::<T>(user.clone());
+        let contract_id = 1;
+
+        // let elapsed_seconds = 200 * 6; // 20 min (200 blocks) later
+        // let stamp: u64 = Timestamp::<T>::get().saturated_into::<u64>() + elapsed_seconds * 1000;
+        // Timestamp::<T>::set_timestamp((stamp).try_into().unwrap());
```

Comment:
remove comments?

Reply:
This commented code is because I don't manage to fix some issue while generating the `weights.rs` even if it works in testing context
So this benchmark would be complete if it takes into account some elapsed time and bill for some cost which is not zero (because `elapsed time = 0 `=> `cost = 0`)

Reply:
this is what I have when I uncomment the code after running this cmd:
~~~sh
./target/release/tfchain benchmark pallet --chain=dev --pallet=pallet_smart_contract --extrinsic="*" --steps=50 --repeat=20 --execution=wasm --heap-pages=409 --output ./pallets/pallet-smart-contract/src/weights.rs --template ./.maintain/frame-weight-template.hbs
~~~

![image](https://github.com/threefoldtech/tfchain/assets/73958772/e48329e9-04f2-4333-8e0d-b7841ae9b5c1)

Reply:
solved!
fix will be follow in next PR
see https://github.com/threefoldtech/tfchain/pull/739/commits/141796e7577160517d0c22d6b02dbf70b625fe6e
sorry for the confusion